File: lib/src/model/notification_details.dart
/// Notification ma'lumotlari - iOS va Android uchun
class NotificationDetails {
  /// Notification channel ID (Android uchun majburiy)
  final String channelId;

  /// Notification sarlavhasi
  final String title;

  /// Notification matni
  final String body;

  /// Unique notification ID - update/cancel uchun
  /// Agar null bo'lsa, avtomatik generate qilinadi
  final int? id;

  /// Custom sound fayl nomi (extension bilan)
  /// iOS: "sound_name.caf" yoki "sound_name.wav"
  /// Android: channel'dan olinadi, bu field ignore qilinadi
  final String? soundUri;

  /// iOS uchun qo'shimcha sarlavha (title va body orasida)
  final String? subtitle;

  /// iOS uchun app icon badge raqami
  final int? badge;

  /// Katta rasm URL yoki local path
  /// Android: Big Picture Style
  /// iOS: Attachment
  final String? imageUrl;

  /// Kichik icon URL yoki local path (masalan, user profile picture)
  /// Android: Large Icon
  /// iOS: Bu field ignore qilinadi
  final String? largeIcon;

  /// Notification rangi (hex format: "#FF5722")
  /// Faqat Android uchun
  final String? color;

  /// Bosilganda avtomatik yo'qolishi
  /// Default: true
  final bool autoCancel;

  /// Ovoz va vibration'siz ko'rsatish
  /// Default: false
  final bool silent;

  /// Custom data - notification bosilganda qayta ishlash uchun
  /// Masalan: {"userId": "123", "type": "message"}
  final Map<String, dynamic>? payload;

  const NotificationDetails({
    required this.channelId,
    required this.title,
    required this.body,
    this.id,
    this.soundUri,
    this.subtitle,
    this.badge,
    this.imageUrl,
    this.largeIcon,
    this.color,
    this.autoCancel = true,
    this.silent = false,
    this.payload,
  });

  /// Map'ga o'tkazish (Native'ga yuborish uchun)
  Map<String, dynamic> toMap() {
    return {
      'channelId': channelId,
      'title': title,
      'body': body,
      'description': body, // Android compatibility
      if (id != null) 'id': id,
      if (soundUri != null) 'soundUri': soundUri,
      if (subtitle != null) 'subtitle': subtitle,
      if (badge != null) 'badge': badge,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (largeIcon != null) 'largeIcon': largeIcon,
      if (color != null) 'color': color,
      'autoCancel': autoCancel,
      'silent': silent,
      if (payload != null) 'payload': payload,
    };
  }

  /// Map'dan yaratish (Native'dan kelganda)
  factory NotificationDetails.fromMap(Map<dynamic, dynamic> map) {
    return NotificationDetails(
      channelId: map['channelId'] as String,
      title: map['title'] as String,
      body: (map['body'] ?? map['description']) as String,
      id: map['id'] as int?,
      soundUri: map['soundUri'] as String?,
      subtitle: map['subtitle'] as String?,
      badge: map['badge'] as int?,
      imageUrl: map['imageUrl'] as String?,
      largeIcon: map['largeIcon'] as String?,
      color: map['color'] as String?,
      autoCancel: map['autoCancel'] as bool? ?? true,
      silent: map['silent'] as bool? ?? false,
      payload: map['payload'] != null
          ? Map<String, dynamic>.from(map['payload'] as Map)
          : null,
    );
  }

  /// Copy with - qisman o'zgartirish uchun
  NotificationDetails copyWith({
    String? channelId,
    String? title,
    String? body,
    int? id,
    String? soundUri,
    String? subtitle,
    int? badge,
    String? imageUrl,
    String? largeIcon,
    String? color,
    bool? autoCancel,
    bool? silent,
    Map<String, dynamic>? payload,
  }) {
    return NotificationDetails(
      channelId: channelId ?? this.channelId,
      title: title ?? this.title,
      body: body ?? this.body,
      id: id ?? this.id,
      soundUri: soundUri ?? this.soundUri,
      subtitle: subtitle ?? this.subtitle,
      badge: badge ?? this.badge,
      imageUrl: imageUrl ?? this.imageUrl,
      largeIcon: largeIcon ?? this.largeIcon,
      color: color ?? this.color,
      autoCancel: autoCancel ?? this.autoCancel,
      silent: silent ?? this.silent,
      payload: payload ?? this.payload,
    );
  }

  @override
  String toString() {
    return 'NotificationDetails('
        'channelId: $channelId, '
        'title: $title, '
        'id: $id, '
        'soundUri: $soundUri'
        ')';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is NotificationDetails &&
        other.channelId == channelId &&
        other.title == title &&
        other.body == body &&
        other.id == id;
  }

  @override
  int get hashCode {
    return Object.hash(channelId, title, body, id);
  }
}

File: additional_files/notification_details.swift

import Foundation

struct NotificationDetails: Codable {
    let channelId: String
    let title: String
    let body: String

    let id: Int?
    let soundUri: String?       // iOS uchun: "sound.caf"
    let subtitle: String?       // iOS uchun
    let badge: Int?             // iOS uchun
    let imageUrl: String?       // Attachment
    let largeIcon: String?      // iOS'da ishlatilmaydi
    let color: String?          // iOS'da ishlatilmaydi
    let autoCancel: Bool?
    let silent: Bool?
    let payload: [String: String]?

    enum CodingKeys: String, CodingKey {
        case channelId
        case title
        case body
        case description        // Android compatibility
        case id
        case soundUri
        case subtitle
        case badge
        case imageUrl
        case largeIcon
        case color
        case autoCancel
        case silent
        case payload
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        channelId = try container.decode(String.self, forKey: .channelId)
        title = try container.decode(String.self, forKey: .title)

        if let bodyValue = try? container.decode(String.self, forKey: .body) {
            body = bodyValue
        } else {
            body = try container.decode(String.self, forKey: .description)
        }

        // Optional fields
        id = try? container.decode(Int.self, forKey: .id)
        soundUri = try? container.decode(String.self, forKey: .soundUri)
        subtitle = try? container.decode(String.self, forKey: .subtitle)
        badge = try? container.decode(Int.self, forKey: .badge)
        imageUrl = try? container.decode(String.self, forKey: .imageUrl)
        largeIcon = try? container.decode(String.self, forKey: .largeIcon)
        color = try? container.decode(String.self, forKey: .color)
        autoCancel = try? container.decode(Bool.self, forKey: .autoCancel)
        silent = try? container.decode(Bool.self, forKey: .silent)
        payload = try? container.decode([String: String].self, forKey: .payload)
    }
    
    init(
        channelId: String,
        title: String,
        body: String,
        id: Int? = nil,
        soundUri: String? = nil,
        subtitle: String? = nil,
        badge: Int? = nil,
        imageUrl: String? = nil,
        largeIcon: String? = nil,
        color: String? = nil,
        autoCancel: Bool? = nil,
        silent: Bool? = nil,
        payload: [String: String]? = nil
    ) {
        self.channelId = channelId
        self.title = title
        self.body = body
        self.id = id
        self.soundUri = soundUri
        self.subtitle = subtitle
        self.badge = badge
        self.imageUrl = imageUrl
        self.largeIcon = largeIcon
        self.color = color
        self.autoCancel = autoCancel
        self.silent = silent
        self.payload = payload
    }
    
    // MARK: - Encodable
        func encode(to encoder: Encoder) throws {
            var container = encoder.container(keyedBy: CodingKeys.self)

            try container.encode(channelId, forKey: .channelId)
            try container.encode(title, forKey: .title)
            // body’ni **body** kaliti bilan yozamiz (description emas)
            try container.encode(body, forKey: .body)

            try container.encodeIfPresent(id, forKey: .id)
            try container.encodeIfPresent(soundUri, forKey: .soundUri)
            try container.encodeIfPresent(subtitle, forKey: .subtitle)
            try container.encodeIfPresent(badge, forKey: .badge)
            try container.encodeIfPresent(imageUrl, forKey: .imageUrl)
            try container.encodeIfPresent(largeIcon, forKey: .largeIcon)
            try container.encodeIfPresent(color, forKey: .color)
            try container.encodeIfPresent(autoCancel, forKey: .autoCancel)
            try container.encodeIfPresent(silent, forKey: .silent)
            try container.encodeIfPresent(payload, forKey: .payload)
            // Eslatma: .description kalitini umuman encode qilmaymiz.
        }

    /// Map'dan yaratish (MethodCall arguments'dan)
    static func fromMap(_ map: [String: Any]) throws -> NotificationDetails {
        let data = try JSONSerialization.data(withJSONObject: map)
        return try JSONDecoder().decode(NotificationDetails.self, from: data)
    }

    /// Map'ga o'tkazish
    func toMap() -> [String: Any?] {
        var result: [String: Any?] = [
            "channelId": channelId,
            "title": title,
            "body": body,
        ]

        if let id = id { result["id"] = id }
        if let soundUri = soundUri { result["soundUri"] = soundUri }
        if let subtitle = subtitle { result["subtitle"] = subtitle }
        if let badge = badge { result["badge"] = badge }
        if let imageUrl = imageUrl { result["imageUrl"] = imageUrl }
        if let largeIcon = largeIcon { result["largeIcon"] = largeIcon }
        if let color = color { result["color"] = color }
        if let autoCancel = autoCancel { result["autoCancel"] = autoCancel }
        if let silent = silent { result["silent"] = silent }
        if let payload = payload {
            result["payload"] = payload.mapValues { $0 }
        }

        return result
    }
}
